tools/tcat_ble_client/cli/tlv_commands.py

```diff
@@ -0,0 +1,78 @@
+"""
+  Copyright (c) 2024, The OpenThread Authors.
+  All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are met:
+  1. Redistributions of source code must retain the above copyright
+     notice, this list of conditions and the following disclaimer.
+  2. Redistributions in binary form must reproduce the above copyright
+     notice, this list of conditions and the following disclaimer in the
+     documentation and/or other materials provided with the distribution.
+  3. Neither the name of the copyright holder nor the
+     names of its contributors may be used to endorse or promote products
+     derived from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+  POSSIBILITY OF SUCH DAMAGE.
+"""
+
+from .base_commands import BleCommand, DataNotPrepared, CommandResultTLV
+from ble.ble_stream_secure import BleStreamSecure
+from enum import StrEnum
+from tlv.tlv import TLV
+from tlv.tcat_tlv import TcatTLVType
+
+
+class TlvTreeCommands(StrEnum):
+    LIST = "list"
+    SEND = "send"
+
+
+class TlvCommand(BleCommand):
+
+    def get_log_string(self) -> str:
+        return 'Sending custom TLV message...'
+
+    def get_help_string(self) -> str:
+        return 'Send custom TLV message.'
+
+    async def execute_default(self, args, context):
```

Comment:
It was made due to the 'tlv list' command, its purpose is only to list available TLV types and not send anything to the TCAT device. The default implementation assumes that the command always sends something to the device:

```
try:
    data = self.prepare_data(args, context)
    response = await bless.send_with_resp(data)
    if not response:
        return
    tlv_response = TLV.from_bytes(response)
    return CommandResultTLV(tlv_response)
```

In effect when calling `tlv list` we get an unexpected error message: `ERROR:__main__:a bytes-like object is required, not 'NoneType'`

To avoid this I've moved the `response = await bless.send_with_resp(data)` under the `if data:` statement.
In the meantime, I found that you made changes to the default implementation. I will apply them to the `TlvCommand` command implementation.

Reply:
done

Reply:
Still kind of inconsistent. The original reason why the `if  data` wasn't there is because prepare data shiuld throw an exception if data werent prepared correctly.  From what I can see that would do the trick here as well.

Reply:
After your comment, I spent some time finding why you would like me to raise an exception in `prepare data` because then it will be caught by:
```
        except DataNotPrepared as err:
            print('Command failed', err)
```
and still print the `Command failed` at the end in the `tlv list` output, which could be misleading to the end user.

After a while, I found that I had taken the wrong approach to implement the `tlv` command tree. The first I noticed was that there are two command classes 'BleCommand' and its parent 'Command'. And the second was that there is a `self._subcommands' field.

My problem was that I wanted to use the `BleCommand` to implement a single command class with subcommands without setting the `self._subcommands`.

I've just pushed a new implementation, please take a look at it. I hope that now it will be alright :).